Improved colormaps support

almarklein opened the pull request:
Closes #104.

* [x] Proper support for colormaps that are gray, gray+alpha, rgb, rgba.
* [x] Per-vertex colors can be gray, gray+alpha, rgb, rgba.
* For Image:
  * [x] proper support for gray, gray+alpha, rgb, rgba.
  * [x] Colormaps.
* For Volume and VolumeSlice:
  * [x] proper support for gray, gray+alpha, rgb, rgba.
  * [x] Colormaps
* For Mesh:
  * [x] Refactor colormaps.
  * [x] Remove clim property. 
  * [x] Per-vertex colors.
* For MeshSlice:
   * [x] Added note about not implementing colormapping or per-vertex colors for now.
* For Line (and ThinLine):
  * [x] Colormaps.
  * [x] Per-vertex colors (was already implemented).
* For points:
  * [x] Colormaps.
  * [x] Per-vertex colors (was already implemented).
* Do I forget objects, maybe background?
* [ ] A few standard colormaps.
* [ ] A complete lib of builtin colormaps.
* [x] Docs.
* [x] Self review / check todos.
* [x] Rebase and update validation example to be tested on CI.

I'm building this list up for completeness, not everything will go in this PR.

----
<img width="872" alt="image" src="https://user-images.githubusercontent.com/3015475/151766054-fe3bb90e-e804-426a-881b-16347a276f86.png">

<img width="882" alt="image" src="https://user-images.githubusercontent.com/3015475/151766148-8fd7ff81-b319-42c3-b5bf-31c308938b1c.png">

<img width="865" alt="image" src="https://user-images.githubusercontent.com/3015475/151766195-b0533b0b-6a5f-4a0b-b45e-554de5b6bbf1.png">

<img width="880" alt="image" src="https://user-images.githubusercontent.com/3015475/151766310-8123a917-1420-4d6b-9739-d884e1df0149.png">

Korijn commented on examples/validate_image_colormap.py:
```diff
@@ -0,0 +1,35 @@
+
+
+if __name__ == "__main__":
+    canvas.request_draw(lambda: renderer.render(scene, camera))
```
This line needs to be executed on import for the test to pass, otherwise no draw function is registered :)

almarklein commented on pygfx/renderers/wgpu/_shadercomposer.py:
```diff
@@ -606,8 +608,60 @@ def common_functions(self):
+                let color = vec4<f32>(color_value.rrr, 1.0);
+            $$ elif colormap_nchannels == 2
+                let color = vec4<f32>(color_value.rrr, color_value.g);
+            $$ elif colormap_nchannels == 3
+                let color = vec4<f32>(color_value.rgb, 1.0);
+            $$ else
+                let color = vec4<f32>(color_value.rgb, color_value.a);
+            $$ endif
+            return color;
```
This shadercode (only marked the last few lines) is the colormap shader code used by all materials.

almarklein commented on pygfx/renderers/wgpu/imagerender.py:
```diff
@@ -11,61 +11,35 @@
-            return color_value;
-        }
-    """
+def handle_colormap(geometry, material, shader):
```
The Image renderer has a `handle_colormap` that is also used by the Volume.

almarklein commented on pygfx/renderers/wgpu/imagerender.py:
```diff
@@ -140,6 +122,94 @@ def image_renderer(render_info):
     ]
 
 
+sampled_value_to_color = """
+        fn sampled_value_to_color(value_rgba: vec4<f32>) -> vec4<f32> {
```
Also used by Volume.

almarklein commented on pygfx/renderers/wgpu/pointsrender.py:
```diff
@@ -2,8 +2,49 @@
+from ...resources import Texture, TextureView
+
+
+def handle_colormap(geometry, material, shader):
```
This `handle_colormap` function (in pointsrender.py`) is also used in the mesh and line renderers.

almarklein commented:
Let's add two-or-so colormaps in a next PR.

Korijn commented on pygfx/geometries/_base.py:
```diff
@@ -29,7 +29,8 @@ class Geometry(ResourceContainer):
     * ``texcoords``: Texture coordinates used to lookup the color for a vertex.
       Can be Nx1, Nx2 or Nx3, corresponding to a 1D, 2D and 3D texture map.
-    * ``colors``: Nx4 per-vertex colors (rgba).
+    * ``colors``: Per-vertex) colors. Must be NxM, with M 1-4 for gray,
```
I think there's a bracket typo here

Korijn commented:
Pretty awesome, I take it you've constructed examples to test all the different cases and confirmed they work? If so, LGTM!

almarklein commented:
Yes there's `colormap_image.py` and `colormap_mesh.py`. The latter can easily be tweaked to render points and lines instead. I added screenshots to the top of this PR. I also added one validation example. I should probably add one or two more. Will also do that in a new pr.
